File: cloud-backend/app/schemas/patient.py
from pydantic import BaseModel, EmailStr, Field
from datetime import datetime
from typing import Optional

class PatientCreate(BaseModel):
    """
    INPUT model - what mobile app sends when registering a patient
    """
    Name: str = Field(..., min_length=1, description="Patient's full name")
    Age: int = Field(..., gt=0, le=120, description="Patient's age in years")
    Gender: str = Field(..., description="male, female, or other")
    ContactNumber: str = Field(..., min_length=10, description="Phone number")
    Email: EmailStr = Field(..., description="Patient's email")
    # removing pregnency:bool because a women cannot be always a pregnent
    Notes: Optional[str] = Field(None, description="Optional clinical notes")
    

    class Config:
        json_schema_extra={
            "example":{ 
                "Name":"Viraj Guranna Yalawar",
                "Age":45,
                "Gender": "male",
                "ContactNumber": "+91-9845497179",
                "Email": "virajgyalawar05@gmail.com",
                "Notes": "Regular checkup"
            }
        }

class PatientResponse(BaseModel):
    """
    OUTPUT model - what API returns after registering patient
    """
    Id: str = Field(alias="_id", default="")
    PatientId: str = Field(alias="patient_id")  # Unique ID for this patient
    Name: str
    Age: int
    Gender: str
    ContactNumber: str
    Email: str
    Notes: Optional[str] = None
    CreatedAt: datetime
    OwnerId: str  # Who registered/owns this patient
    
    class Config:
        populate_by_name = True

File: mobile-app/lib/domain/entities/patient.dart
class Patient {
  final String id;
  final String name;
  final int age;
  final String gender;
  final String contactNumber;
  final String email;
  final String? notes;
  final DateTime createdAt;

  Patient({
    required this.id,
    required this.name,
    required this.age,
    required this.gender,
    required this.contactNumber,
    required this.email,
    this.notes,
    required this.createdAt,
  });

  Patient copyWith({
    String? id,
    String? name,
    int? age,
    String? gender,
    String? contactNumber,
    String? email,
    String? notes,
    DateTime? createdAt,
  }) {
    return Patient(
      id: id ?? this.id,
      name: name ?? this.name,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      contactNumber: contactNumber ?? this.contactNumber,
      email: email ?? this.email,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'PatientId': id,
      'name': name,
      'Name': name,
      'age': age,
      'Age': age,
      'gender': gender,
      'Gender': gender,
      'contactNumber': contactNumber,
      'ContactNumber': contactNumber,
      'email': email,
      'Email': email,
      'notes': notes,
      'Notes': notes,
      'createdAt': createdAt.toIso8601String(),
      'CreatedAt': createdAt.toIso8601String(),
    };
  }

  factory Patient.fromJson(Map<String, dynamic> json) {
    return Patient(
      id: (json['PatientId'] ??
              json['patient_id'] ??
              json['patientId'] ??
              json['id'] ??
              json['_id'] ??
              '')
          .toString(),
      name: (json['Name'] ?? json['name'] ?? '').toString(),
      age: ((json['Age'] ?? json['age'] ?? 0) as num).toInt(),
      gender: (json['Gender'] ?? json['gender'] ?? '').toString(),
      contactNumber:
          (json['ContactNumber'] ?? json['contactNumber'] ?? json['contact_number'] ?? '')
              .toString(),
      email: (json['Email'] ?? json['email'] ?? '').toString(),
      notes: (json['Notes'] ?? json['notes'])?.toString(),
      createdAt: json['CreatedAt'] != null
          ? DateTime.parse(json['CreatedAt'].toString())
          : json['createdAt'] != null
              ? DateTime.parse(json['createdAt'].toString())
              : DateTime.now(),
    );
  }

  @override
  String toString() => 'Patient(id: $id, name: $name, age: $age)';
}
